src/h264rtpdepacketizer.cpp

```diff
@@ -0,0 +1,138 @@
+ */
+
+#if RTC_ENABLE_MEDIA
+
+#include "h264rtpdepacketizer.hpp"
+#include "track.hpp"
+
+#include "impl/logcounter.hpp"
+
+#include <cmath>
+#include <utility>
+
+#ifdef _WIN32
+#include <winsock2.h>
+#else
+#include <arpa/inet.h>
+#endif
+
+namespace rtc {
+
+const unsigned long stapaHeaderSize = 1;
+const auto fuaHeaderSize = 2;
+const auto rtpHeaderSize = 12;
+
+const auto naluTypeBitmask = std::byte(0x1F);
+const auto naluTypeSTAPA = std::byte(24);
+const auto naluTypeFUA = std::byte(28);
+const auto fuaEndBitmask = std::byte(0x40);
+const auto naluRefIdcBitmask = std::byte(0x60);
+
+message_vector H264RtpDepacketizer::buildFrame(message_vector::iterator first,
+                                               message_vector::iterator last) {
+	message_vector out = {};
+	auto fua_buffer = std::vector<std::byte>{};
+
+	for (auto it = first; (it - 1) != last; it++) {
+		auto pkt = it->get();
+		auto pktParsed = reinterpret_cast<const rtc::RtpHeader *>(pkt->data());
+		auto headerSize =
+		    rtpHeaderSize + pktParsed->csrcCount() + pktParsed->getExtensionHeaderSize();
+		auto naluType = pkt->at(headerSize) & naluTypeBitmask;
+
+		if (fua_buffer.size() != 0 || naluType == naluTypeFUA) {
+			if (fua_buffer.size() == 0) {
+				fua_buffer.push_back(std::byte(0));
+			}
+
+			std::copy(pkt->begin() + headerSize + fuaHeaderSize, pkt->end(),
+			          std::back_inserter(fua_buffer));
+
+			if ((pkt->at(headerSize + 1) & fuaEndBitmask) != std::byte(0)) {
+				auto naluRefIdc = pkt->at(headerSize) & naluRefIdcBitmask;
+				auto fragmentedNaluType = pkt->at(headerSize + 1) & std::byte(naluTypeBitmask);
+				fua_buffer.at(0) = naluRefIdc | fragmentedNaluType;
+
+				out.push_back(make_message(fua_buffer));
+				fua_buffer.clear();
+			}
+		} else if (naluType > std::byte(0) && naluType < std::byte(24)) {
+			out.push_back(make_message(pkt->begin() + headerSize, pkt->end()));
+		} else if (naluType == naluTypeSTAPA) {
+			auto currOffset = stapaHeaderSize + headerSize;
+
+			while (currOffset < pkt->size()) {
+				auto naluSize =
+				    uint16_t(pkt->at(currOffset)) << 8 | uint8_t(pkt->at(currOffset + 1));
+
+				currOffset += 2;
+
+				if (pkt->size() < currOffset + naluSize) {
+					throw std::runtime_error("STAP-A declared size is larger then buffer");
+				}
+
+				out.push_back(
+				    make_message(pkt->begin() + currOffset, pkt->begin() + currOffset + naluSize));
+				currOffset += naluSize;
+			}
+
+		} else {
+			throw std::runtime_error("Unknown H264 RTP Packetization");
+		}
+	}
+
+	return out;
+}
+
+void H264RtpDepacketizer::incoming(message_vector &messages, const message_callback &) {
+	for (auto message : messages) {
+		if (message->type != Message::Binary) {
+			continue;
+		}
+
+		if (message->size() < sizeof(RtpHeader)) {
+			PLOG_VERBOSE << "RTP packet is too small, size=" << message->size();
+			continue;
+		}
+
+		this->rtp_buffer.push_back(make_message(message->begin(), message->end()));
```

Comment:
Fixed

Reply:
Doesn't `mRtpBuffer` need a copy because it will last across reads?

When I don't copy I get this crash

```
Thread 7 "RTC worker" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff47fa6c0 (LWP 1950035)]
rtc::RtpHeader::timestamp (this=0x0) at /home/sean/workspaces/libdatachannel/src/rtp.cpp:58
58      uint32_t RtpHeader::timestamp() const { return ntohl(_timestamp); }
(gdb) bt
#0  rtc::RtpHeader::timestamp (this=0x0) at /home/sean/workspaces/libdatachannel/src/rtp.cpp:58
#1  0x00007ffff7d38c05 in rtc::H264RtpDepacketizer::incoming(std::vector<std::shared_ptr<rtc::Message>, std::allocator<std::shared_ptr<rtc::Message> > >&, std::function<void (std::shared_ptr<rtc::Message>)> const&) (this=0x476790, messages=std::vector of length 1, capacity 1 = {...}) at /home/sean/workspaces/libdatachannel/src/h264rtpdepacketizer.cpp:116
#2  0x00007ffff7dba38c in rtc::impl::Track::incoming (this=this@entry=0x477540, message=std::shared_ptr<rtc::Message> (empty) = {...})
    at /home/sean/workspaces/libdatachannel/src/impl/track.cpp:146
#3  0x00007ffff7d8da6b in rtc::impl::PeerConnection::dispatchMedia (this=0x4251f0, message=std::shared_ptr<rtc::Message> (use count 6, weak count 0) = {...})
    at /home/sean/workspaces/libdatachannel/src/impl/peerconnection.cpp:599
```

Reply:
Messages are held as `shared_ptr` in order to keep them until the last reference is removed, so there should be no need to create a copy (also, copying this way loses message metadata like type).

The issue here is that the message is cleared or consumed by a move for some reason, so `size()` is 0 and `data()` ends up returning `nullptr`. `Track` should be the only place were a message is consumed to move it to the user as binary. It might indicate a deeper problem like a message leaking to the user.